[12.0][FIX] website_sale_product_minimal_price: Take into account all pricelist item

CarlosRoca13 opened the pull request:
Change the get_combination_info_pricelist_atributes method to get the
correct scale of prices.

cc @Tecnativa TT29708

please @pedrobaeza @chienandalu review it

CarlosRoca13 committed:
[FIX] website_sale_product_minimal_price: Take into account all pricelist items

Previous approach only considered pricelist items in the same pricelist as the current
one, but you can have some items that point to another pricelists, so we are changing
the approach to get all possible items, no matter the pricelist, get the minimal quantity
which they apply to, and then get the price for that quantity to see if with the price
computation, they return a different value, marking them in the scale table. With this,
we get a compromise between performance and accuracy.

NOTE: There are still left intentionally items for "All products" that can have other
scaled prices, but as this can have a lot of impact in performance and it's improbable
that someone configures pricelists this way, it's better to not include them.

TT29708

pedrobaeza commented:
This is not `[IMP]`, but `[FIX]`, as it wasn't working previously. The commit message can be: 

```
[FIX] website_sale_product_minimal_price: Take into account all pricelist items

Previous approach only considered pricelist items in the same pricelist as the current 
one, but you can have some items that point to another pricelists, so we are changing
the approach to get all possible items, no matter the pricelist, get the minimal quantity
which they apply to, and then get the price for that quantity to see if with the price
computation, they return a different value, marking them in the scale table. With this,
we get a compromise between performance and accuracy.

NOTE: There are still left intentionally items for "All products" that can have other
scaled prices, but as this can have a lot of impact in performance and it's improbable
that someone configures pricelists this way, it's better to not include them.
```

pedrobaeza reviewed (approved):
Code and functional review, taking into account my commit message comment

CarlosRoca13 commented:
@pedrobaeza changed

pedrobaeza commented:
Don't put `@Tecnativa` on the commit message. You can put the TT, but not the mention.

CarlosRoca13 commented:
@pedrobaeza changed

chienandalu commented on website_sale_product_minimal_price/controllers/main.py:
```diff
@@ -16,48 +16,33 @@ def get_combination_info_pricelist_atributes(
+            ("categ_id", "in", list(map(int, product.categ_id.parent_path.split('/')[0:-1]))),
+            ("min_quantity", ">", 0),
+        ]).mapped("min_quantity")
+        qty_list = sorted(set(qty_list))
```
How do you know that the pricelist is chained to the used one?

CarlosRoca13 replied:
In this case we take all the pricelist items asociated to this product to know what are the `minimal_qty`s that have to be checked, so we will know the scaled prices. This method is used just on the grid product price scale

CarlosRoca13 committed:
[FIX] website_sale_product_minimal_price: No compute prices
When the product has a sales price = 0, his variants has no extra price and a pricelist depends on other pricelists to compute the prices, the algorithm takes a random variant.

Doing this changes we take care all pricelists that depends on the main to compute the prices asociated to the product.

TT30398

pedrobaeza commented:
@chienandalu can you confirm?

pedrobaeza commented:
/ocabot merge patch

pedrobaeza commented:
@CarlosRoca13 please take note for fw-porting the patch to v13 when possible.
